feat: on low wasm memory hook (#286)
dfinity/interface-spec: spec/index.md

-   The canister may export functions with name `canister_composite_query <name>` and type `() -> ()`.

> -   The canister may export functions with name `canister_on_low_wasm_memory` and type `() -> ()`.

Reviewer: ```suggestion
-   The canister may export a function with the name `canister_on_low_wasm_memory` and type `() -> ()`.
```

:::

#### On Low Wasm Memory {#on-low-wasm-memory}

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 2: @mraszyk WDYT? I think that more important information is how much memory is left. Furthermore, if we decide to check the replica's remaining memory, we cannot use the upper bound on the wasm memory size for that.

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer: Makes sense to me. Should the inequality be <= or < ?

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 2: "<" makes more sense, since the default is 0, hence if we have a default value of the treshold the hook will never be invoked. I updated sugesstion.

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 3: Discussion in the interface spec meeting:
- The semantics here are unclear, for instance in the case that the canister has a certain memory allocation and uses both heap and stable memory. What does "the canister's remaining wasm memory size" then refer to, exactly?
- The formulation seems to indicate a guarantee that actually cannot be provided (without an explicit memory allocation): Suppose some canister makes a big memory allocation, which brings some 10s of thousands of canisters on the same subnet over/under the threshold. We would have to call the hook on all of those 10s of thousands of canisters, but it's very unlikely that all of those would be able to satisfy their memory needs.

Thus, we are not convinced that this change of semantics is beneficial, given that the only way for a canister to guarantee that it can still allocate memory (from the shared resource of overall subnet memory) is via memory allocation, which however has ambiguous semantics as described above.

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 3: Another complication here is that the canister may not be able to _pay for_ allocating additional memory due to the cycles prepayment when the subnet memory becomes scarce. Should that trigger the condition as well?

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 2: I agree it is better to look only at the `canister's memory`, not the `subnet available memory`. In that case:

1. If the certain application uses both heap and stable memory then the `remaining wasm memory` will be:

min(`memory_allocation` - `used_stable_memory`, 4 GB) - `used_heap_memory`

2. If we look only at the canister's memory, the allocation of the other canister cannot trigger the hook on the current canister, hence that will not be a problem. Furthermore, by checking if the `remaining wasm memory` is below the `threshold` we will ensure enough `wasm memory` in case of usage of `memory allocation` (first part in the formula above), which cannot be done if the hook is checking if the `wasm memory` if above the `threshold`.

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 3: The formula would likely have to include a few more terms that are accounted for in `memory_allocation`, such as the wasm module size or the size of the canister history. When that is done, it seems an acceptable API.

I guess in my opinion there is no clearly "better" API. The "old" one is a bit easier to understand in that it uses more basic values, but has the disadvantage of probably requiring the canister to change the threshold at least upon significant allocation of additional stable memory. This "new" variant may be a bit easier to use, but is bit more complicated to understand and would need to be adapted when new features (e.g. multiple wasm memories) are supported.

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

Reviewer 2: In that case, I will go forward and update the specs to use “remaining memory drops to below a threshold”.

The threshold `t` can be defined in the [canister's settings](#ic-canister_status) and by default it is set to 2<sup>64</sup> − 1.

:::note

While the above function is scheduled immediately once the condition above is triggered, it may not necessarily be executed immediately if it does not have enough cycles.

> If the canister gets frozen immediately after the function is scheduled for execution, the function will run once the canister's unfrozen _if_ the canister's wasm memory remains above the threshold `t`.

Reviewer: So you don't plan to treat it analogously to call_on_cleanup after all?

Author: We discussed it and preferred to move forward with the scheduling approach. The reason for that is that `call_on_cleanup`, unlike this hook, cannot make outbound calls. Making outbound calls in the hook while also treating it like `call_on_cleanup` causes unnecessary complications in the implementation.

Reviewer: > The reason for that is that call_on_cleanup, unlike this hook, cannot make outbound calls.

As long as it can schedule a one-off timer, I don't see a major issue with this limitation.

:::

#### On Low Wasm Memory {#on-low-wasm-memory}

A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's wasm memory size in bytes grows from below a threshold `t` to >= `t`.

> The threshold `t` can be defined in the [canister's settings](#ic-canister_status) and by default it is set to 2<sup>64</sup> − 1.

Reviewer: The default value should be 0 with your above change, right?

Reviewer 2: That is correct, thank you!

> The threshold `t` can be defined in the [canister's settings](#ic-canister_status) and by default it is set to 2<sup>64</sup> − 1.

Reviewer 2: ```suggestion
The threshold `t` can be defined in the [canister's settings](#ic-canister_status) and by default it is set to 0.
```

:::

#### On Low Wasm Memory {#on-low-wasm-memory}

> A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's remaining wasm memory size in bytes falls from above a threshold `t` to < `t`.

Reviewer: ```suggestion
A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's remaining wasm memory size in bytes falls from above a threshold `t` to strictly less than `t`.
```

:::

#### On Low Wasm Memory {#on-low-wasm-memory}

A canister can export a function with the name `canister_on_low_wasm_memory`, which is scheduled whenever the canister's remaining wasm memory size in bytes falls from above a threshold `t` to strictly less than `t`.
The threshold `t` can be defined in the [canister's settings](#ic-canister_status) and by default it is set to 0.

:::note

While the above function is scheduled immediately once the condition above is triggered, it may not necessarily be executed immediately if it does not have enough cycles.

> If the canister gets frozen immediately after the function is scheduled for execution, the function will run once the canister's unfrozen _if_ the canister's wasm memory remains above the threshold `t`.

Reviewer: > If the canister gets frozen immediately after the function is scheduled for execution, the function will run once the canister's unfrozen _if_ the canister's wasm memory remains above the threshold `t`.

This property is not established by the implementation.

-   `F`: from `canister_inspect_message`

> -   `T`: from *system task* (`canister_heartbeat` or `canister_global_timer` `canister_on_low_wasm_memory`)

Reviewer 2: ```suggestion
-   `T`: from *system task* (`canister_heartbeat` or `canister_global_timer` or `canister_on_low_wasm_memory`)
```

-   `F`: from `canister_inspect_message`

> -   `T`: from *system task* (`canister_heartbeat` or `canister_global_timer` or `canister_on_low_wasm_memory`)

Reviewer 3: If we make this of type `T`, then a canister could update the threshold in a setting and allocate a new page, which should make the hook trigger again. So if we guarantee *immediate* execution, one can build a chain of calls to the hook. So we need to make sure that *immediate* may mean that it executes as part of the next execution round (just that nothing else runs in between).

```

*Call context creation: On low wasm memory*

> If canister `C` exports a method with name `canister_on_low_wasm_memory` and `S.on_low_wasm_memory_hook_status[C]` is `Ready`, the IC will create the corresponding call context and set `S.on_low_wasm_memory_hook_status[C]` to `Executed`.

Reviewer: ```suggestion
If `S.on_low_wasm_memory_hook_status[C]` is `Ready` for a canister `C`, the IC will create the corresponding call context and set `S.on_low_wasm_memory_hook_status[C]` to `Executed`.
```
